### DISCUSSION TITLE
[META] Design discussion

### DISCUSSION THREAD

**Discussion by matmair** [inventree/InvenTree#4226] [General]
As there is interest in helping with InvenTrees UI / UX I would be interested to hear feedback from users.
1. How is the current design suiting your workflow
2. How do you interact the most with Inventree? The web interface, the app or an API
3. What interactions/interfaces could be designed better?

This is meant as a (gentle) sounding board to get some ideas on how the community feels about the overall UX. There are no promises attached regarding a redesign etc.
If you really want to see a specific change you can always open a PR or start an issue.

**Comment by matmair** [comment]
I am going to start with my current feelings as of 0.9.2:
1. The overall design works out to be only a few clicks once I scan a part - that is very convenient for me
2. I mainly use the web interface if I interact with it. Although most days it is just my automation doing its thing
3. 2 things stand out right now:
3a. Some UI elements scale badly on tablet screens - I am using InvenTree with iPads and some stuff just overflows the page on detail pages
3b. I would love to be able to use actions on tables with a right click or long tab - just like a menu in Windows

**Comment by eeintech** [comment]
Here's my initial thoughts:
1. I use the search... a lot. That's my way to access a part quickly. I rarely, rarely use the categories tree.
2. The only times I use the category tree is for parametric table: I go to the right category and filter parts by parameters I'm interested into. I do the same thing as I do on Digi-Key or any other supplier website, filters are super important in my opinion.
3. I rarely hit the "Stock" or "Build" tabs, as they are both accessible via the part view. The only time I go there is for setting up a stock location or build order.
4. Same goes with "Buy" or "Sell" tabs, I only used them for **setup** then management is done directly on the part view.
5. So I spend 90% of my time on a part view, and 10% on setup and parametric tables.

My opinion for a design revamp:
- Make the search central to the UI/UX, maybe even introduce a quick shortcut like CTRL+K on some documentation webpages
- Improve very much the parametric table and functions, those are essential when looking on a supplier website therefore translate 100% to InvenTree
- Separate setup vs management functions: setup does not need to be made so prominent in the UI, but could be accessed through a secondary view on the "management" page

**Comment by miggland** [comment]
I think I've been using InvenTree for about 6 months now. It's helping a lot, and it's obviously evolving! Our processes, because of major changes to the company, are also evolving fast.

1. Very well. Mostly no issues. The ideas below are because you asked, I haven't put enough work/interest in yet to make the issues :)
2. 90% Web interface, 9.5% via custom scripts for import/export using inventree-python, 0.5% app(*)
3. Some ideas, where I see improvements (and I haven't compared these to any issues):
       - To go into a certain part category, I have to switch to the category view - and then switch back for the part list.
       - Parameter tables perform badly at the moment. However, I don't use them that much at the moment.
       - I haven't understood the pricing feature yet, but I see a lot of potential here. I know it's being/has been reviewed lately.
       - I often want to add some extra variables to models. One example is having the size and weight of a shipment saved, so that I can access it later on. Right now, I use the notes field. I realise that InvenTree can't be adapted specifically to all of these, so some more flexibility would be great. I think the Metadata option is the basis for this!

(*) Especially my warehouse guy should be using the app more in the future. Things like marking items as received and packing Sales Orders

**Reply by eeintech** [reply]
> 3\. I often want to add some extra variables to models. One example is having the size and weight of a shipment saved, so that I can access it later on. Right now, I use the notes field. I realise that InvenTree can't be adapted specifically to all of these, so some more flexibility would be great. I think the Metadata option is the basis for this!

Maybe "parameters" should apply to all models in InvenTree :thinking: I can see it apply to PO,  SO, BO, Stock items, locations, etc.
Or a more generic "metadata" class right.

> To go into a certain part category, I have to switch to the category view - and then switch back for the part list.

This used to not be the case :disappointed: I agree it should be improved.

> Parameter tables perform badly at the moment.

+1

**Reply by miggland** [reply]
> Maybe "parameters" should apply to all models in InvenTree thinking I can see it apply to PO, SO, BO, Stock items, locations, etc. Or a more generic "metadata" class right.

If you have more input on this, glad to hear it here:
https://github.com/inventree/InvenTree/issues/4202

**Reply by eeintech** [reply]
Oh so metadatas are in already on most models but not exposed in the UI, right?

**Reply by miggland** [reply]
They're in a few (see the ones checked in the other issue), but not in most. None are exposed in the UI.

**Reply by matmair** [reply]
Metadata was meant to be used by plugins for data that is handled totally by them. So if you do not want to use AppMixin and custom migrations you can use metadata to extend data models and permanently save data.

My thoughts:
* Metadata is for backend/plugins
* Parameter should be extended to include datatypes, validation, a nice and performant UI, parametric search etc. They could be extended to orders, stock and more. Parameter would be for users that do not want to code something up

Thoughts @miggland @eeintech

**Reply by miggland** [reply]
In any case, I think all/most models should have metadata, so that they are accessible for plugins.

The other idea came later - if the metadata model is there, one option would be to make a "Custom data" plugin which saves its data as metadata.

> ... to include datatypes, validation, a nice and performant UI, parametric search etc.
Yes, ideally that would be the case.

> They could be extended to orders, stock and more
Yes, as with metadata, should apply to most/all models.

> Parameter would be for users that do not want to code something up
Definitely. They should definitely be accessible via API.

My suggestion was along the lines of using what is available, but also grew over time (originally, I was thinking of a specific plugin for my use case, and thus metadata..). Maybe it's not the best solution.

The current parameter system, which is there for Parts, could be extended as well I guess. Currently, also manufacturer parts have parameters - but these work differently.
